test/Pester/EditorServices.Integration.Tests.ps1

```diff
@@ -0,0 +1,86 @@
+
+$script:ExceptionRegex = [regex]::new('\s*Exception: (.*)$', 'Compiled,Multiline,IgnoreCase')
+function ReportLogErrors
+{
+    param(
+        [Parameter()][string]$LogPath,
+
+        [Parameter()][ref]<#[int]#>$FromIndex = 0,
+
+        [Parameter()][string[]]$IgnoreException = @()
+    )
+
+    $logEntries = Parse-PsesLog $LogPath |
+        Where-Object Index -ge $FromIndex.Value
+
+    # Update the index to the latest in the log
+    $FromIndex.Value = ($FromIndex.Value,$errorLogs.Index | Measure-Object -Maximum).Maximum
+
+    $errorLogs = $logEntries |
+        Where-Object LogLevel -eq Error |
+        Where-Object {
+            $match = $script:ExceptionRegex.Match($_.Message.Data)
+
+            (-not $match) -or ($match.Groups[1].Value.Trim() -notin $IgnoreException)
+        }
+
+    if ($errorLogs)
+    {
+        $errorLogs | ForEach-Object { Write-Error "ERROR from PSES log: $($_.Message.Data)" }
+    }
+}
+
+Describe "Loading and running PowerShellEditorServices" {
+    BeforeAll {
+        Import-Module -Force "$PSScriptRoot/../../module/PowerShellEditorServices"
+        Import-Module -Force "$PSScriptRoot/../../tools/PsesPsClient/out/PsesPsClient"
+        Import-Module -Force "$PSScriptRoot/../../tools/PsesLogAnalyzer"
+
+        $logIdx = 0
+        $psesServer = Start-PsesServer
+        $client = Connect-PsesServer -PipeName $psesServer.SessionDetails.languageServicePipeName
+    }
+
+    # This test MUST be first
+    It "Starts and responds to an initialization request" {
+        $request = Send-LspInitializeRequest -Client $client
+        $response = Get-LspResponse -Client $client -Id $request.Id
+        $response.Id | Should -BeExactly $request.Id
+
+        ReportLogErrors -LogPath $psesServer.LogPath -FromIndex ([ref]$logIdx)
+    }
+
+    # This test MUST be last
```

Comment:
I don't think this is the right approach. I think that there should be 2 `Context`'s:

1. "Message handling"
2. "Startup and shutdown"

So the tests would be structured like this:

```powershell
Describe "Loading and running PowerShellEditorServices" {
    BeforeAll {
        Import-Module -Force "$PSScriptRoot/../../module/PowerShellEditorServices"
        Import-Module -Force "$PSScriptRoot/../../tools/PsesPsClient/out/PsesPsClient"
        Import-Module -Force "$PSScriptRoot/../../tools/PsesLogAnalyzer"

        $logIdx = 0
    }
    Context "Startup and shutdown" {
        It "Can start up and shutdown" {
            # Start up
            $psesServer = Start-PsesServer
            $client = Connect-PsesServer -PipeName $psesServer.SessionDetails.languageServicePipeName

            # Shutdown
            $request = Send-LspShutdownRequest -Client $client
            $response = Get-LspResponse -Client $client -Id $request.Id
            $response.Id | Should -BeExactly $request.Id
            $response.Result | Should -BeNull
            # TODO: The server seems to stay up waiting for the debug connection
            # $psesServer.PsesProcess.HasExited | Should -BeTrue
           
            try
            {
                $psesServer.PsesProcess.Kill()
            }
            finally
            {
                try
                {
                    $psesServer.PsesProcess.Dispose()
                }
                finally
                {
                    $client.Dispose()
                }
            }

            ReportLogErrors -LogPath $psesServer.LogPath -FromIndex ([ref]$logIdx)
        }
    }

    Context "LSP messages" {
        BeforeAll {
            # Start up
            $psesServer = Start-PsesServer
            $client = Connect-PsesServer -PipeName $psesServer.SessionDetails.languageServicePipeName
        }
        AfterAll {
            try
            {
                $psesServer.PsesProcess.Kill()
            }
            finally
            {
                try
                {
                    $psesServer.PsesProcess.Dispose()
                }
                finally
                {
                    $client.Dispose()
                }
            }
        }
        It "Can send initialize request" {
            $request = Send-LspInitializeRequest -Client $client
            $response = Get-LspResponse -Client $client -Id $request.Id
            $response.Id | Should -BeExactly $request.Id

            ReportLogErrors -LogPath $psesServer.LogPath -FromIndex ([ref]$logIdx)
        }
        #
        # As time goes on we add more It's here for the other messages
        #
    }
}
```

Reply:
That way the order doesn't matter and we can easily add new tests for other messages.

For the future:
I think the `Send-LspInitializeRequest` will actually probably need to be moved to the BeforeAll eventually once we add other messages that depend on that one having gone through.

We'll get there. Gotta figure out which messages depend on what and align the tests as much as possible to that back and forth.

Reply:
Yeah I don't think there's any particularly obvious approach available at the moment. I think given that we're not planning impending development on this module that we should merge what we have since it works and is less code and then continue as is.

Reply:
Fair

Reply:
Definitely don't disagree with you though -- "unit" testing is very bad at sequential logic with dependent states.

I think we should refactor as you say, but will defer that to another PR.